jobs/kpi-forecasting/kpi_forecasting/models/funnel_forecast.py

```diff
@@ -311,58 +389,53 @@ def _auto_tuning(self, segment_settings: SegmentModelSettings) -> Dict[str, floa
             for v in itertools.product(*segment_settings.grid_parameters.values())
         ]
 
-        test_dat = self._build_model_dataframe(
-            segment_settings, "train", add_log_growth_cols
+        test_dat = self._build_train_dataframe(
+            observed_df, segment_settings, add_log_growth_cols
         )
         bias = []
 
         for params in param_grid:
             m = self._build_model(segment_settings, params)
             m.fit(test_dat)
 
-            df_cv = cross_validation(m, **segment_settings.cv_settings)
-
-            df_bias = df_cv.groupby("cutoff")[["yhat", "y"]].sum().reset_index()
-            df_bias["pcnt_bias"] = df_bias["yhat"] / df_bias["y"] - 1
-            # Prophet splits the historical data when doing cross validation using
-            # cutoffs. The `.tail(3)` limits the periods we consider for the best
-            # parameters to the 3 most recent cutoff periods.
-            bias.append(df_bias.tail(3)["pcnt_bias"].mean())
+            crossval_metric = self._get_crossvalidation_metric(
+                m, segment_settings.cv_settings
+            )
+            bias.append(crossval_metric)
 
         min_abs_bias_index = np.argmin(np.abs(bias))
 
         return param_grid[min_abs_bias_index]
 
-    def _add_regressors(self, dat: pd.DataFrame, regressors: List[ProphetRegressor]):
+    def _add_regressors(self, df: pd.DataFrame, regressors: List[ProphetRegressor]):
         """
         Add regressor columns to the dataframe for training or prediction.
 
         Args:
-            dat (pd.DataFrame): The input dataframe.
+            df (pd.DataFrame): The input dataframe.
             regressors (List[ProphetRegressor]): The list of regressors to add.
 
         Returns:
             pd.DataFrame: The dataframe with regressors added.
         """
-        df = dat.copy().rename(columns=self.column_names_map)
-        df["ds"] = pd.to_datetime(df["ds"])
         for regressor in regressors:
             regressor = self._fill_regressor_dates(regressor)
             # finds rows where date is in regressor date ranges and sets that regressor
             ## value to 1, else 0
-            df[regressor.name] = np.where(
-                (df["ds"] >= pd.to_datetime(regressor.start_date))
-                & (df["ds"] <= pd.to_datetime(regressor.end_date)),
-                0,
-                1,
-            )
+            df[regressor.name] = (
+                (df["ds"] >= pd.to_datetime(regressor.start_date).date())
+                & (df["ds"] <= pd.to_datetime(regressor.end_date).date())
+            ).astype(int)
         return df
 
-    def _predict(self, segment_settings: SegmentModelSettings) -> pd.DataFrame:
+    def _predict(
+        self, dates_to_predict_raw: pd.DataFrame, segment_settings: SegmentModelSettings
```

Comment:
Was the purpose of passing the `dates_to_predict` df as an attribute to make testing simpler?

Reply:
Yeah that's the main reason